packages/core/src/typeschema/crawler.ts

```diff
@@ -0,0 +1,113 @@
+import { Resource } from '@medplum/fhirtypes';
+import { getTypedPropertyValue, toTypedValue } from '../fhirpath';
+import { TypedValue } from '../types';
+import { arrayify, isLowerCase } from '../utils';
+import { getDataType, InternalTypeSchema } from './types';
+
+export interface ResourceVisitor {
+  onEnterObject?: (path: string, value: TypedValue, schema: InternalTypeSchema) => void;
+  onExitObject?: (path: string, value: TypedValue, schema: InternalTypeSchema) => void;
+  visitProperty?: (
+    parent: TypedValue,
+    key: string,
+    path: string,
+    propertyValues: (TypedValue | TypedValue[] | undefined)[],
+    schema: InternalTypeSchema
+  ) => void;
+}
+
+export function crawlResource(
+  resource: Resource,
+  visitor: ResourceVisitor,
+  schema?: InternalTypeSchema,
+  initialPath?: string
+): void {
+  new ResourceCrawler(resource, visitor, schema, initialPath).crawl();
+}
+
+class ResourceCrawler {
+  private readonly rootResource: Resource;
+  private readonly visitor: ResourceVisitor;
+  private readonly schema: InternalTypeSchema;
+  private readonly initialPath: string;
+
+  constructor(rootResource: Resource, visitor: ResourceVisitor, schema?: InternalTypeSchema, initialPath?: string) {
+    this.rootResource = rootResource;
+    this.visitor = visitor;
+
+    if (schema) {
+      this.schema = schema;
+    } else {
+      this.schema = getDataType(rootResource.resourceType);
+    }
+
+    if (initialPath) {
+      this.initialPath = initialPath;
+    } else {
+      this.initialPath = rootResource.resourceType;
+    }
+  }
+
+  crawl(): void {
+    this.crawlObject(toTypedValue(this.rootResource), this.schema, this.initialPath);
+  }
+
+  private crawlObject(obj: TypedValue, schema: InternalTypeSchema, path: string): void {
+    if (this.visitor.onEnterObject) {
+      this.visitor.onEnterObject(path, obj, schema);
+    }
```

Comment:
Should this handle the resource type check itself, and become e.g. `onEnterResource`?